### PR TITLE
Creation Quote Metric

### PR DESCRIPTION
**Pull request by mtakeda**
#### What problem is this solving?

Sending quotation created metrics to analytics in order to have more business data to take decisions and analyze user's behavior.

#### How to test it?

Create a quote in `https://takeda--b2bstore005.myvtex.com/deals`
After one hour (maximum), data should appear in Analytics Redshift, into table `"vtex"."schemaless"."b2b_suite_buyerorg_data_raw"`

[Workspace](https://takeda--b2bstore005.myvtex.com/deals)

#### Screenshots or example usage:
No changes in layout. Current quotation create workflow kept as is.
image

Redshit query and result example:
```
SELECT
    payload.kind,
    payload.account,
    payload.fields.quote_id,
    payload.fields.quote_reference_name,
    payload.fields.creation_date,
    payload.fields.send_to_sales_rep,
    payload.fields.org_id,
    payload.fields.cost_center_id,
    payload.fields.cost_center_name,
    payload.fields.buy_org_id,
    payload.fields.buy_org_name,
    payload.fields.member_id,
    payload.fields.member_email,
    payload.fields.role
FROM
    "vtex"."schemaless"."b2b_suite_buyerorg_data_raw"
WHERE payload.name = 'b2b-suite-buyerorg-data'
order by ingestion_time desc
```
image

type:
```
type Metric = {
  name: 'b2b-suite-buyerorg-data'
  kind: 'create-quote-ui-event'
  description: 'Create Quotation Action - UI'
  account: string
}

type QuoteFieldsMetric = {
  org_id: string
  cost_center_id: string
  cost_center_name: string
  buy_org_id: string
  buy_org_name: string
  member_id: string
  member_email: string
  role: string
  creation_date: string
  quote_id: string
  quote_reference_name: string
  send_to_sales_rep: boolean
}

type QuoteMetric = Metric & { fields: QuoteFieldsMetric }
```

#### Describe alternatives you've considered, if any.
NA

#### Related to / Depends on
NA

### CODE REVIEW

File: docs/README.md
Comment (mairatma) [inline docs/README.md:254]: I love that you documented this, but these docs automatically show up to our end-users (from the merchant or partner agencies), so they're not supposed to have internal information for our VTEX devs, only information for devs implementing this solution for their stores.

Cléber mentioned that we'll have an interna-docs repo to hold this kind of doc, so we should move this there.

File: docs/README.md
Comment (mairatma) [inline docs/README.md:255]: I think you forgot these lines

File: react/components/QuoteDetails/QuoteDetails.tsx
Comment (mairatma) [inline react/components/QuoteDetails/QuoteDetails.tsx:133]: This will make us fetch all that data every time this page is rendered, even if we end up not needing it.

It seems like we only ever need this when sending the metric, right? So it's best if we only do this data fetching at that time, within the `sendMetric` function.

In that case you won't be doing the fetching using hooks (hooks are these functions with names starting with `use`). React hooks have logic behind the scenes that cause the screen to be rerendered when the data is loading, and then again when it arrives. For your use case we should do everything behind the scenes and without causing rerenders. If these data fetches were not graphql calls you could just use `axios` for that, for example. In graphql's case you can use a simple library like [graphql-request](https://www.npmjs.com/package/graphql-request) instead.

File: react/components/QuoteDetails/QuoteDetails.tsx
Comment (mairatma) [inline react/components/QuoteDetails/QuoteDetails.tsx:132]: One of the advantages of graphql is that you don't need to do separate fetches to get information. Since you need all three of these to send the metric, you can build a single graphql query joining them all together in a single request. Also, it's probably better to build a new query for this use case anyway, since you probably don't need the exact same data that those existing queries were asking for. You can drop the unused fields and join the queries together, resulting in a better API request.

File: react/components/QuoteDetails/QuoteDetails.tsx
Comment (mairatma) [inline react/components/QuoteDetails/QuoteDetails.tsx:226]: I'm not 100% sure,  but it's possible that all this data is already present in the session or segment cookie, so it would be good to double check before making the changes I suggested to the data fetching.

File: react/utils/metrics.ts
Comment (mtakeda) [reply react/utils/metrics.ts:91]: In this case, I'm not sure I need to catch errors. The idea of not having an await is not having to wait for the response from analytics API and not delay a bit more the response for users.

Comment (mairatma) [reply react/utils/metrics.ts:144]: I agree that it's not very important for you to catch errors here, but in that case you should remove the try/catch. Without the await it won't work well.

If you want to keep catching the errors here you can add the await. That won't really change the behavior of the call, because you don't await the `sendMetric` call itself, so it will still run in the background without affecting the rendering thread in any way.

File: react/utils/metrics.ts
Comment (mairatma) [inline react/utils/metrics.ts:75]: Just a small indentation fix:

```suggestion
      getUserByEmail(email: $email) @context(provider: "vtex.storefront-permissions") {
        costId
      }
```

File: react/utils/metrics.ts
Comment (mairatma) [inline react/utils/metrics.ts:117]: In line 98 the code is assuming that `namespaces.profile.email.value` is always set, but here you're handling the case where `namespaces` and `profile` might not be set. If that's a possibility we need to also add this to lines 97 and 98, skipping the whole metric process in case the values are not set (since those are required).

File: react/utils/metrics.ts
Comment (mairatma) [inline react/utils/metrics.ts:83]: When a GraphQL request fails, the response status is often 200 anyway, depending on the type of the error (mainly because there can be partial errors). So there would be no exception thrown here, but `data` might be undefined, or it might only have part of the information you requested. In those cases the response will contain a field called `errors`, so you can check for that here.

```suggestion
  const { data, errors } = (
    await axios.post(GRAPHQL_URL(accountName, workspace), query)
  ).data

```

More information on graphql errors [here](https://learning.postman.com/open-technologies/blog/graphql-error-handling/).

File: react/utils/metrics.ts
Comment (mairatma) [inline react/utils/metrics.ts:85]: You're using an optional operator here, so you're assuming that `data.getUserByEmail` might not be set, right? If that's a possibility then you shouldn't cast to string here, as it'll make `costId`'s type not consider the use case where it will be `undefined`. Is it ok if `costId` isn't set? If it isn't it's better to bail early here, and if it's ok then you should let it be `string | undefined`.

File: react/components/QuoteDetails/QuoteDetails.tsx
Comment (mairatma) [inline react/components/QuoteDetails/QuoteDetails.tsx:76]: You can also grab `accountName` from here instead of from the session. Maybe the user email too, but that one I'm not sure about.

**Review by mairatma — COMMENTED** [review]
Thanks Mauro! Just smaller suggestions now.

File: react/utils/metrics.ts
Comment (mtakeda) [reply react/utils/metrics.ts:117]: Instead of skipping the while metric process, I'll check the namespace, profile, and email, and let it save null values. I believe it's better to save the metric with the data we have than not save it at all

Comment (mairatma) [reply react/utils/metrics.ts:114]: I agree, if the metric is still useful, let's save it!

**Review by mairatma — DISMISSED** [review]
Lgtm, thanks! 👏

**Comment by mairatma** [comment]
@mtakeda don't forget to select the version bump within vtex-io-ci-cd bot's message at the further above in this PR, or that check will keep failing.

File: manifest.json
Comment (mairatma) [inline manifest.json:4]: This shouldn't be changed here in the PR. It will be changed automatically by the vtex io bot once this PR is merged. You should revert this change, and just select the right type of version bump in the question post from vtex io bot.

File: react/package.json
Comment (mairatma) [inline react/package.json:3]: Same here, please revert this line.

File: react/utils/metrics/createQuote.ts
Comment (mairatma) [inline react/utils/metrics/createQuote.ts:99]: I've just noticed that this PR also has the typos:

```suggestion
      buyer_org_id: metricsData.organization,
      buyer_org_name: metricsData.organizationName,
```

Can you check if there are others and fix them? Same for the metrics for using quotes.

**Review by mairatma — APPROVED** [review]

**Pull request merged**
